Docs to change Container runtime (#30141)
kubernetes/website: content/en/docs/tasks/administer-cluster/migrating-from-dockershim/Change-runtime.md

---

> title: "Migrating Container runtime from dockershim to containerd"

Reviewer: this seems to be an opinionated suggestion for transition.
what if the user wishes to use cri-o instead of containerd?

in the CR popularity survey we did before docker was first, followed by containerd and then cri-o.
so if containerd wins here by being second, it makes sense, but cri-o might not be very happy about this.

cc @saschagrunert for feedback on CRI.

Reviewer 2: Thanks for reaching out! I think the whole guide is opinionated to a particular setup, which is probably fine. Mentioning CRI-O as alternative container runtimes to containerd would make us very happy, though. :upside_down_face:

Reviewer: ok, i think it's worth denoting with a sentence near the top of the guide that this covers an example scenario for migrating from dockershim to containerd and that alternative container runtime such as cri-o can be picked from the https://kubernetes.io/docs/setup/production-environment/container-runtimes/ page.

Reviewer 3: That sounds like a good way to head off issues around https://kubernetes.io/docs/contribute/style/content-guide/#third-party-content

We make an exception for container runtimes - these _are_ necessary out-of-project packages that you must have to make a working Kubernetes cluster - but we don't like to be in the business of picking favorites.

weight: 10
content_type: task 
---

- Drain Node
```
# replace <node-to-drain> with the name of your node you are draining
kubectl drain <node-to-drain> --ignore-daemonsets
```

> - Stop docker

Reviewer 3: ```suggestion
- Stop the Docker daemon
```

```
systemctl stop kubelet
systemctl stop docker
```
- Install & start containerd 

> ### For linux systems

Reviewer 3: nit: Linux (capital L)

Use the following commands to install Containerd on your system:

> Install and configure prerequisites:

Reviewer: the guide seems to be duplicating steps that already existing the in the "installing a container runtime" page:
https://kubernetes.io/docs/setup/production-environment/container-runtimes/
instead we can cross-link.

the steps can be generalized as:
- drain node
- stop kubelet
- uninstall old CR
- install new CR (CR of your chose, not picking a preference in this page, cross-link to the main CR page)
- update kubelet config and kubeadm bits (optional)
- uncordon node
- check if it works

i know that it works and i've seen users already do it, but before we merge such a guide for hotswap SIG node must tell us if this is even supported and if something unpredictable can happen. cgroup (drivers)? proper cleanup after uninstalling the old CR - logs, container state, etc?

cc @SergeyKanzhelev

Reviewer 3: I think we do want a “dummies guide to” here, because the Docker deprecation is already worrying quite a few people.

```shell
cat <<EOF | sudo tee /etc/modules-load.d/containerd.conf
overlay
br_netfilter
EOF

sudo modprobe overlay
sudo modprobe br_netfilter

# Setup required sysctl params, these persist across reboots.
cat <<EOF | sudo tee /etc/sysctl.d/99-kubernetes-cri.conf
net.bridge.bridge-nf-call-iptables  = 1
net.ipv4.ip_forward                 = 1
net.bridge.bridge-nf-call-ip6tables = 1
EOF

# Apply sysctl params without reboot
sudo sysctl --system
```

Install containerd:

> 1. Install the `containerd.io` package from the official Docker repositories. 

Reviewer 3: Why use the Docker repository for this? I would recommend that people fetch it from their
OS distribution or, failing that, from source and build it.

I suspect that telling people to get containerd from Docker is liable to add to the confusion.

> Instructions for setting up the Docker repository for your respective Linux distribution and 
> installing the `containerd.io` package can be found at 
> [Install Docker Engine](https://docs.docker.com/engine/install/#server).

Reviewer 3: ```suggestion
   You can find instructions for installing `containerd` at
   [Starting Containerd](https://containerd.io/docs/getting-started/#starting-containerd).
```

2. Configure containerd:

   ```shell
   sudo mkdir -p /etc/containerd
   containerd config default | sudo tee /etc/containerd/config.toml
   ```

3. Restart containerd:

   ```shell
   sudo systemctl restart containerd
   ```

> ### For windows powershell

Reviewer 3: Consider splitting this into two task pages, one for Linux nodes and one for Windows nodes.

Start a Powershell session, set `$Version` to the desired version (ex: `$Version=1.4.3`), 
and then run the following commands:

1. Download containerd:

   ```powershell
   curl.exe -L https://github.com/containerd/containerd/releases/download/v$Version/containerd-$Version-windows-amd64.tar.gz -o containerd-windows-amd64.tar.gz
   tar.exe xvf .\containerd-windows-amd64.tar.gz
   ```

2. Extract and configure:

   ```powershell
   Copy-Item -Path ".\bin\" -Destination "$Env:ProgramFiles\containerd" -Recurse -Force
   cd $Env:ProgramFiles\containerd\
   .\containerd.exe config default | Out-File config.toml -Encoding ascii

   # Review the configuration. Depending on setup you may want to adjust:
   # - the sandbox_image (Kubernetes pause image)
   # - cni bin_dir and conf_dir locations
   Get-Content config.toml

   # (Optional - but highly recommended) Exclude containerd from Windows Defender Scans
   Add-MpPreference -ExclusionProcess "$Env:ProgramFiles\containerd\containerd.exe"

>    ```

Reviewer 4: - Each command should typically be its own code section (unless relevant commands can be grouped in order for the text description to explain things more effectively) with a description above it and the comments can be put as text and not as comments.
    - Please use [this](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/) for reference.
- `# (Optional - but highly recommended) Exclude containerd from Windows Defender Scans`: Please describe why this is recommended.

Author: This piece of code block has also been taken from the documentation here https://github.com/kubernetes/website/blob/main/content/en/docs/setup/production-environment/container-runtimes.md#container-runtimes, but yes I will add more descriptions to make this more clear and if required separately group them.

3. Start containerd:

   ```powershell
   .\containerd.exe --register-service
   Start-Service containerd
   ```
- configure kubelet to use containerd

> Edit the file `/var/lib/kubelet/kubeadm-flags.env` and add the containerd runtime to the flags. `--container-runtime=remote` and `--container-runtimeendpoint=unix:///run/containerd/containerd.sock"`

Reviewer: `kubeadm-flags.env` should be denoted as kubeadm specific - i.e. "If you are using kubeadm ...".

another kubeadm specific would be the following:
```
Kubeadm stores the CRI socket for each host as an annotation in the Node object for that host.
To change it you must do the following:

- Execute `kubectl edit no <NODE-NAME>` on a machine that has the kubeadm `/etc/kubernetes/admin.conf` file.
This will start a text editor where you can edit the Node object. To choose a text editor you can set the `KUBE_EDITOR` environment variable.
- Change the value of `kubeadm.alpha.kubernetes.io/cri-socket` from `/var/run/dockershim.sock`
to the CRI socket path of your chose (for example `unix:///run/containerd/containerd.sock`). Note that new CRI socket paths must be prefixed with `unix://` ideally.
- Save the changes in the text editor, which will update the Node object.
```

i think we can have the `/var/lib/kubelet/kubeadm-flags.env` and the above in a sub-section called - "If you are using kubeadm"

- restart kubelet
`systemctl start kubelet`

- verify pods are running 

> Run `kubectl get nodes -o wide` and containerd appears as the runtime for the node we just changed.

Reviewer 4: ```
- configure kubelet to use containerd
Edit the file `/var/lib/kubelet/kubeadm-flags.env` and add the containerd runtime to the flags. `--container-runtime=remote` and `--container-runtimeendpoint=unix:///run/containerd/containerd.sock"`
```
This is not rendered on separate lines if that was your original intent. And same for the ones below it as well.

- finally if everything goes well remove docker
```
apt purge docker-ce docker-ce-cli
OR
yum remove docker-ce docker-ce-cli

> ```

Reviewer 4: - Is this still under Windows Powershell?
- Not sure if the `OR` here is the right way to go about it. Maybe a separate code section?

> ```

Reviewer 3: I would use https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs and have a tab for Debian-heritage systems and another for RedHat-heritage systems.

---

> title: "Migrating Container runtime from dockershim to conatiainerd"

Reviewer 5: conatiainerd --> containerd
